tuta-sdk/rust/src/lib.rs

```diff
@@ -0,0 +1,182 @@
+// might or might not work for WASM
+
+// Option 2:
+// metamodel -> Kotlin/Swift classes
+// need to be able to covert from ParsedEntity -> Kotlin/Swift class
+// will generate a bit less code but we need to write the conversion for every platform
+// will work for WASM for sure
+
+impl Display for TypeRef {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "TypeRef({}, {})", self.app, self.type_)
+    }
+}
+
+trait AuthHeadersProvider {
+    /// Gets the HTTP request headers used for authorizing REST requests
+    fn auth_headers(&self, model_version: u32) -> HashMap<String, String>;
+}
+
+/// The authorization status and credentials of the SDK
+enum LoginState {
+    NotLoggedIn,
+    LoggedIn { access_token: String },
+}
+
+/// Contains all the high level mutable state of the SDK
+struct SdkState {
+    login_state: RwLock<LoginState>,
+    client_version: String,
+}
+
+/// The external facing interface used by the consuming code via FFI
+#[derive(uniffi::Object)]
+pub struct Sdk {
+    state: Arc<SdkState>,
+    entity_client: Arc<EntityClient>,
+}
+
+#[uniffi::export]
+impl Sdk {
+    #[uniffi::constructor]
+    pub fn new(base_url: String, rest_client: Arc<dyn RestClient>, client_version: &str) -> Sdk {
+        let type_model_provider = init_type_model_provider();
+        // TODO validate parameters
+        let instance_mapper = Arc::new(InstanceMapper::new(type_model_provider));
+        let state = Arc::new(SdkState {
+            login_state: RwLock::new(LoginState::NotLoggedIn),
+            client_version: client_version.to_owned(),
+        });
+        Sdk {
+            state: state.clone(),
+            entity_client: Arc::new(EntityClient::new(
+                rest_client,
+                instance_mapper,
+                &base_url,
+                state,
+            )),
+        }
+    }
+
+    /// Authorizes the SDK's REST requests via inserting `access_token` into the HTTP headers
+    pub fn login(&self, access_token: &str) {
+        let mut login_state = self.state.login_state.write().unwrap();
+        if let LoginState::LoggedIn { .. } = *login_state {
+            panic!("Already logged in!")
+        }
+        *login_state = LoginState::LoggedIn {
+            access_token: access_token.to_owned(),
+        }
+    }
+
+
+    /// Generates a new interface to operate on mail entities
+    pub fn mail_facade(&self) -> MailFacade {
+        MailFacade::new(self.entity_client.clone())
+    }
+}
+
+impl AuthHeadersProvider for SdkState {
+    fn auth_headers(&self, model_version: u32) -> HashMap<String, String> {
+        let g = self.login_state.read().unwrap();
+        match g.deref() {
+            LoginState::NotLoggedIn => HashMap::new(),
+            LoginState::LoggedIn { access_token } => {
+                HashMap::from([
+                    ("accessToken".to_owned(), access_token.as_str().to_owned()),
+                    ("cv".to_owned(), self.client_version.to_owned()),
```

Comment:
I am not sure auth headers provider should inject model versions, I think `EntityClient` could do that (or any other common underlying network abstraction we introduce)